### PR TITLE
Update to .Net5

### PR DESCRIPTION
**Pull request by taooceros**
.Net 5 Upgrade.
close #228 

- Net 5 has added several improvement to System.Text.Json, and serialization with private getter and setter has been added.

### CODE REVIEW

**Comment by JohnTheGr8** [comment]
@taooceros looking good thus far. 

Remember to also rename the publish profile: `Flow.Launcher/Properties/PublishProfiles/NetCore3.1-SelfContained.pubxml`

**Comment by taooceros** [comment]
> @taooceros looking good thus far.
> 
> Remember to also rename the publish profile: `Flow.Launcher/Properties/PublishProfiles/NetCore3.1-SelfContained.pubxml`

ok

**Comment by JohnTheGr8** [comment]
@taooceros you forgot to check in the renamed publish-profile file.

File: Plugins/Flow.Launcher.Plugin.WebSearch/Flow.Launcher.Plugin.WebSearch.csproj [hunk 1]
```diff
@@ -119,14 +119,6 @@
...
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </Content>
-    <Page Include="SettingsControl.xaml">
-      <Generator>MSBuild:Compile</Generator>
-      <SubType>Designer</SubType>
-    </Page>
-    <Page Include="SearchSourceSetting.xaml">
-      <Generator>MSBuild:Compile</Generator>
-      <SubType>Designer</SubType>
-    </Page>
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.WebSearch/Flow.Launcher.Plugin.WebSearch.csproj:129]: is this removal intended?

File: Plugins/Flow.Launcher.Plugin.Program/Flow.Launcher.Plugin.Program.csproj [hunk 2]
```diff
@@ -109,7 +109,6 @@
   </ItemGroup>
   
   <ItemGroup>
-    <PackageReference Include="Microsoft.Windows.SDK.Contracts" Version="10.0.18362.2005" />
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.Program/Flow.Launcher.Plugin.Program.csproj:112]: is this removal intended?

**Comment by jjw24** [comment]
apologies if you have already done so but can we also test to double check external plugins still works fine with this upgrade (just to make sure)

File: Plugins/Flow.Launcher.Plugin.WebSearch/Flow.Launcher.Plugin.WebSearch.csproj (hunk 1 above)
Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.WebSearch/Flow.Launcher.Plugin.WebSearch.csproj:129]: Yes, it said duplicate page file, since the default page file has included them.

**Comment by taooceros** [comment]
> apologies if you have already done so, can we test to double check external plugins still works fine with this upgrade (i know it's just core to 5, but just to make sure)

Sure, no need for hurry, just put it here for checking.

File: Plugins/Flow.Launcher.Plugin.Program/Flow.Launcher.Plugin.Program.csproj (hunk 2 above)
Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.Program/Flow.Launcher.Plugin.Program.csproj:112]: Sorry, forget to response. Yes, because .Net 5 change the way of using RT. You can see on the TargetFramework of Program plugin is different from simply net5-windows but adding the version number. That replace the nuget package.

**Comment by taooceros** [comment]
> @taooceros you forgot to check in the renamed publish-profile file.

Thank you for reminding!

**Comment by taooceros** [comment]
Updating to .Net 5 and using WinRT seems cause the PackageManager doesn't work in Program plugin. https://github.com/dotnet/runtime/issues/45632
Not sure why that throw.

**Comment by jjw24** [comment]
is this still draft or ready for review

**Comment by taooceros** [comment]
This should be ready now

**Comment by jjw24** [comment]
Could you please list all the tests that have been done so we can cover all basis

**Comment by taooceros** [comment]
> Could you please list all the tests that have been done so we can cover all basis

Will do that after I have my computer. 😉

**Comment by taooceros** [comment]
Test Checked:
1. All internal plugins work as intended. 
    - Bug for `new PackageManager()` hasn't been fixed, but it only happens on the first time, so a try catch can fix it.
2. External plugins work as intended, even when their Target Framework is netcoreapp 3.1
3. Python plugin works as intended.
4. Have used it for a while and haven't encountered other bugs.

**Comment by taooceros** [comment]
Another things is that .Net 5 obsolete BinaryFormatter because of security issue (though I think it will be secure for us). Should we find an alternative for that?

**Comment by jjw24** [comment]
> Another things is that .Net 5 obsolete BinaryFormatter because of security issue (though I think it will be secure for us). Should we find an alternative for that?

Where is it used?

**Comment by taooceros** [comment]
> 
> > Another things is that .Net 5 obsolete BinaryFormatter because of security issue (though I think it will be secure for us). Should we find an alternative for that?
> 
> Where is it used?

Only in a few place about the cache. Win32 and UWP, and image usage storage.

**Comment by jjw24** [comment]
if we can change it, lets do it. Are there any preferred alternatives?

**Comment by taooceros** [comment]
> if we can change it, lets do it. Are there any preferred alternatives?

I am not quite sure. Maybe it is quite possible to simply use JsonStorage, but not sure about whether it will be slower. Another alternative is ProtoBuf-net, and I have taken a try, v2 works fine while v3 doesn't work because of the cyclic reference for UWP.

**Comment by jjw24** [comment]
Remember to update readme as well

**Comment by jjw24** [comment]
can we reference the actual file so people can download straight away

**Comment by taooceros** [comment]
> can we reference the actual file so people can download straight away

ok

**Comment by jjw24** [comment]
>External plugins work as intended, even when their Target Framework is netcoreapp 3.1

Just double checking, this means you have used and tested an external plugin with the upgraded .Net 5 Plugin api

**Comment by taooceros** [comment]
> >External plugins work as intended, even when their Target Framework is netcoreapp 3.1
> 
> Just double checking, this means you have used and tested an external plugin with the upgraded .Net 5 Plugin api

Yes, I have several installed external Plugins and they work fine.

File: Plugins/Flow.Launcher.Plugin.BrowserBookmark/Flow.Launcher.Plugin.BrowserBookmark.csproj [hunk 3]
```diff
@@ -54,22 +55,10 @@
   </ItemGroup>
 
   <ItemGroup>
-    <Content Include="Languages\*.xaml">
-      <Generator>MSBuild:Compile</Generator>
-      <SubType>Designer</SubType>
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </Content>
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.BrowserBookmark/Flow.Launcher.Plugin.BrowserBookmark.csproj:61]: why are we removing this, does bookmark plugin not have any language files?

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Flow.Launcher.Plugin.BrowserBookmark.csproj:61]: Because xaml is included by default for wpf program

Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Flow.Launcher.Plugin.BrowserBookmark.csproj:61]: cool thanks for letting me know. How come we dont do this for other projects?

File: README.md [hunk 4]
```diff
@@ -78,11 +78,11 @@ Yes please, submit an issue to let us know.
 
 ## Developing/Debugging
 
-Flow Launcher's target framework is .Net Core 3.1
+Flow Launcher's target framework is .Net Core 5
```
Comment (jjw24) [inline README.md:81]: .Net 5, no Core any more

File: README.md [hunk 5]
```diff
@@ -78,11 +78,11 @@ Yes please, submit an issue to let us know.
...
 
 Install Visual Studio 2019
 
-Install .Net Core 3.1 SDK via Visual Studio installer or manually from [here](https://dotnet.microsoft.com/download/dotnet-core/thank-you/sdk-3.1.201-windows-x64-installer)
+Install .Net Core 5 SDK via Visual Studio installer or manually from [here](https://dotnet.microsoft.com/download/dotnet/thank-you/sdk-5.0.103-windows-x64-installer)
```
Comment (jjw24) [inline README.md:85]: same as above

File: Plugins/Flow.Launcher.Plugin.BrowserBookmark/Flow.Launcher.Plugin.BrowserBookmark.csproj (hunk 3 above)
Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Flow.Launcher.Plugin.BrowserBookmark.csproj:61]: Because these two are error when building🤣

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Flow.Launcher.Plugin.BrowserBookmark.csproj:61]: Some of the plugin doesn't have a setting page, so doesn't have to enable wpf

File: README.md (hunk 4 above)
Comment (taooceros) [reply README.md:81]: ok will change that

File: Plugins/Flow.Launcher.Plugin.BrowserBookmark/Flow.Launcher.Plugin.BrowserBookmark.csproj (hunk 3 above)
Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Flow.Launcher.Plugin.BrowserBookmark.csproj:61]: I think a couple more plugins do have settings and use wpf, could you make the change so it's consistent while you are making the others please

**Comment by jjw24** [comment]
> Bug for new PackageManager() hasn't been fixed, but it only happens on the first time, so a try catch can fix it.

is this a .Net 5 bug or something we havent noticed in Core

**Comment by taooceros** [comment]
> > Bug for new PackageManager() hasn't been fixed, but it only happens on the first time, so a try catch can fix it.
> 
> is this a .Net 5 bug or something we havent noticed in Core

I think it is bug for CS/WinRT, which in .net core we don't use it.

File: Plugins/Flow.Launcher.Plugin.BrowserBookmark/Flow.Launcher.Plugin.BrowserBookmark.csproj (hunk 3 above)
Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Flow.Launcher.Plugin.BrowserBookmark.csproj:61]: hold on, are you sure it's included by default?
image

**Comment by jjw24** [comment]
hmm Everything plugin is not happy:
image

I suspect it's because of the infra dll too old?

**Comment by taooceros** [comment]
> hmm Everything plugin is not happy:
> image
> 
> I suspect it's because of the infra dll too old?

I think that's for the whole 1.8.0 because of the migration of ISavable.

**Comment by taooceros** [comment]
> > Bug for new PackageManager() hasn't been fixed, but it only happens on the first time, so a try catch can fix it.
> 
> is this a .Net 5 bug or something we havent noticed in Core

I have contacted cswinrt, and they said that they have fixed it, but it still exist though.

**Comment by jjw24** [comment]
ok so we need to sort Everything out before 1.8.0 goes out then right?

**Comment by taooceros** [comment]
> ok so we need to sort Everything out before 1.8.0 goes out then right?

yes, because Everything plugin use an old Flow.Infracturature, but that hasn't been loaded to Flow because Flow exist a new one. Actually, due to the new plugin api changed, everything plugin maybe able to get rid of it.

File: Plugins/Flow.Launcher.Plugin.BrowserBookmark/Flow.Launcher.Plugin.BrowserBookmark.csproj (hunk 3 above)
Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Flow.Launcher.Plugin.BrowserBookmark.csproj:61]: sure

**Comment by taooceros** [comment]
Let me take a look on whether the xaml do have loaded by default.

**Comment by jjw24** [comment]
bookmark plugin got the same issue...

**Comment by taooceros** [comment]
> bookmark plugin got the same issue...

Will take a look on both. Maybe resource xaml is not included by default, but page xaml does😅

**Comment by jjw24** [comment]
I tested adding .Net 5 Flow.Launcher.Plugin dll to WindowWalker plugin project which is .Net Core, getting this error:
image

**Comment by taooceros** [comment]
> I tested adding .Net 5 Flow.Launcher.Plugin dll to WindowWalker plugin project which is .Net Core, getting this error:
> image

Yeah they cannot add it since it is higher version. However, if they keep the current Flow.Launcher.Plugin.dll, they can be loaded successfully by flow.

**Comment by taooceros** [comment]
> > bookmark plugin got the same issue...
> 
> Will take a look on both. Maybe resource xaml is not included by default, but page xaml does😅

I add the Content back, and they work. Thank you for the discovery!

**Comment by jjw24** [comment]
>  if they keep the current Flow.Launcher.Plugin.dll, they can be loaded successfully by flow.

but that will mean these plugins will not get the new updates to Plugin project

**Comment by taooceros** [comment]
> > if they keep the current Flow.Launcher.Plugin.dll, they can be loaded successfully by flow.
> 
> but that will mean these plugins will not get the new updates to Plugin project

If they want to get the updated version, they just need to upgrade to .Net 5

**Comment by jjw24** [comment]
Plugin project will need to be a major version bump then

**Comment by taooceros** [comment]
> Plugin project will need to be a major version bump then

Isn't we bump that from 1.4.0 to 1.5.0? Or you mean to bump it to 2.0.0

**Comment by jjw24** [comment]
bump to 2.0.0, it's a breaking change for plugins

**Comment by taooceros** [comment]
Would you please do that for me? I am not quite sure how to bump version😂

**Comment by jjw24** [comment]
> Would you please do that for me? I am not quite sure how to bump version😂

yeah no problem, leave it with me. I will do that when we release

**Comment by jjw24** [comment]
are these formatting differences because you are using VSCode?
image

**Comment by taooceros** [comment]
> are these formatting differences because you are using VSCode?
> image

I think so.

**Comment by taooceros** [comment]
Well my VS will format it like this ..... 
image

**Comment by jjw24** [comment]
maybe when you have time have a look at how we can sync the format. I dont want to end up with too many 'fix format' commits 😛  

When I get some time I will see if VSCode is a better tool than Visual Studio and switch if so

**Comment by taooceros** [comment]
> maybe when you have time have a look at how we can sync the format. I dont want to end up with too many 'fix format' commits 😛

Yeah sure🤣🤣

> When I get some time I will see if VSCode is a better tool than Visual Studio and switch if so, because i use it for other programming stuff.


I think VSCode is quite great for us to use, even though it will not be able to us to track CPU usage and Memory usage with VSCode.
Omnisharp provides similar programing experience with VS, except that it cannot give intellisence for unusing namespace.
Omnisharp will need similar time to load project as VS. However, it won't block the main thread so that if we only want to edit a few files, it would be really fast for us to do that. 
For VS, we have to wait for the long startup time.... Generally I won't let VS open until I use it, so VSCode makes me feel really great when I need to only edit one or two line of code comparing to VS.

**Comment by jjw24** [comment]
> except that it cannot give intellisence for unusing namespace (so I have to manually using System.Linq before using Select, Where, etc.).

No wonder occasionally you leave unused namespaces around 😆 

PR LGTM, couple more testing and will merge in with 1.8.0

**Review by jjw24 — APPROVED** [review]
